subsys/infinispan/src/main/java/org/commonjava/indy/subsys/infinispan/CacheProducer.java

```diff
@@ -135,28 +167,57 @@ public void start()
     }
 
     /**
-     * Retrieve a cache with a pre-defined configuration (from infinispan.xml) or using the default cache config.
+     * Get a BasicCache instance. If the remote cache is enabled, it will match the named with remote.patterns.
+     * If matched, it will create/return a RemoteCache. If not matched, an embedded cache will be created/returned to the caller.
      */
-    public synchronized <K,V> CacheHandle<K,V> getCache(String named, Class<K> keyClass, Class<V> valueClass )
+    public synchronized <K, V> BasicCacheHandle<K, V> getBasicCache( String named, Class<K> keyClass, Class<V> valueClass )
     {
-        if ( cacheManager == null )
+        BasicCacheHandle<K, V> handle = caches.get( named );
+        if ( handle != null )
         {
-            throw new IllegalStateException( "Cannot access CacheManager. Indy seems to be in a state of shutdown." );
+            return handle;
         }
 
-        CacheHandle<K, V> handle = caches.get( named );
+        if ( remoteConfiguration.isEnabled() && remoteConfiguration.isRemoteCache( named ) )
+        {
+            RemoteCache<K, V> cache = remoteCacheManager.getCache( named );
+            if ( cache == null )
+            {
+                logger.warn( "Remote cache not found, name: {}", named );
+            }
+            else
+            {
+                handle = new RemoteCacheHandle( named, cache, metricsManager, getCacheMetricPrefix( named ) );
+            }
+        }
         if ( handle == null )
         {
-            Cache<K, V> cache = cacheManager.getCache( named );
-
-            final String cacheMetricPrefix = metricsManager == null ?
-                    null :
-                    getSupername( metricsConfig.getNodePrefix(), INDY_METRIC_ISPN, named );
+            logger.debug( "Get embedded cache, name: {}", named );
+            handle = getCache( named, keyClass, valueClass );
+        }
+        caches.put( named, handle );
+        
+        return handle;
+    }
 
-            handle = new CacheHandle( named, cache, metricsManager, cacheMetricPrefix );
+    /**
+     * Retrieve an embedded cache with a pre-defined configuration (from infinispan.xml) or using the default cache configuration.
+     */
+    public synchronized <K, V> CacheHandle<K, V> getCache( String named, Class<K> keyClass, Class<V> valueClass )
+    {
+        BasicCacheHandle<K, V> handle = caches.get( named );
+        if ( handle == null )
```

Comment:
Would something like a Map.calculateIfAbsent() call work more elegantly here? I think this might be old logic...

Reply:
Ok. Will look at it.

Reply:
A new commit is posted including this - "Remove redundant params from getCache/getBasicCache and apply computeIfAbsent"